### PR TITLE
PIPELINE-3900: Fix missing gaps after reprocessing range-detected gaps

### PR DESCRIPTION
**Pull request by tomaslink**
## Problem

When processing a date range, closed gaps detected by **ProcessGroup** or **ProcessBoundaries** were not accompanied by an open version (v1), unlike daily processing which naturally creates one as an intermediate step.

This caused gaps to be **permanently lost** when a subsequent reprocess started after the OFF message but before the ON message: the delete query removed the closed gap, and no open v1 existed as a seed to reconstruct it.

## Root Cause

In daily processing, the pipeline evaluates each day independently. If a vessel goes dark long enough, an open gap is created on the OFF day (or a subsequent day with no messages), and closed the next time a message arrives. This guarantees that an open v1 always exists in the table before the closed v2 is written.

In range processing, the pipeline can see both the OFF and ON messages within the same run, producing a closed gap directly — skipping the open v1 entirely.

## Fix

For each closed gap detected during range processing, evaluate `eval_open_gap` against every day between the OFF and ON messages (excluding the ON day). If the condition is met on any of those days — mirroring what daily processing would have done — emit an open v1 alongside the closed v2.

This ensures range processing produces the same table state as daily processing, making gap reconstruction after any reprocess always possible.

## Testing

Added 4 test cases covering the two components and two edge cases for the OFF message timestamp:

- `recreate_gap_after_reprocess__boundaries_off_far_from_midnight`
- `recreate_gap_after_reprocess__boundaries_off_close_to_midnight`
- `recreate_gap_after_reprocess__group_off_far_from_midnight`
- `recreate_gap_after_reprocess__group_off_close_to_midnight`

---
https://globalfishingwatch.atlassian.net/browse/PIPELINE-3900

### CODE REVIEW

File: src/pipe_gaps/pipelines/detect/fns/process_group.py [hunk 1]
```diff
@@ -70,6 +71,24 @@ def process(
...
+            off_date = datetime_from_timestamp(off_m[self.KEY_TIMESTAMP]).date()
+
+            should_emit_open = any(
+                self._gd.eval_open_gap(off_m, off_date + timedelta(days=i))
+                for i in range(ceil(gap["duration_h"] / 24) - 1)  # exclude the ON day
+            )
+
+            if should_emit_open:
+                logger.debug("Emitting open gap for recovery...")
+                open_gap = self._gd.create_gap(off_m=off_m, gap_id=gap[self.KEY_GAP_ID])
```
Comment (Chr96er) [inline src/pipe_gaps/pipelines/detect/fns/process_group.py]: This seems to fail because the previous positions values aren't set (see e.g. [this Dataflow job](https://console.cloud.google.com/dataflow/jobs/us-central1/2026-05-05_15_44_40-15393614251209670543;logsSeverity=ERROR;graphView=0?project=world-fishing-827&pageState=(%22dfTime%22:(%22l%22:%22dfJobMaxTime%22)))).

Here's the fix claude suggests:

```
open_gap = self._gd.create_gap(                                                                                                                                                                           
      off_m=off_m,
      gap_id=gap[self.KEY_GAP_ID],                                                                                                                                                                            
      base_gap={
          GapDetector.KEY_HOURS_BEFORE: gap[GapDetector.KEY_HOURS_BEFORE],                                                                                                                                    
          GapDetector.KEY_HOURS_BEFORE_TER: gap[GapDetector.KEY_HOURS_BEFORE_TER],                                                                                                                            
          GapDetector.KEY_HOURS_BEFORE_SAT: gap[GapDetector.KEY_HOURS_BEFORE_SAT],                                                                                                                            
          GapDetector.KEY_HOURS_BEFORE_DYN: gap[GapDetector.KEY_HOURS_BEFORE_DYN],                                                                                                                            
      },                                                                                                                                                                                                      
  )
```

**Review by Chr96er — CHANGES_REQUESTED** [review]

File: src/pipe_gaps/pipelines/detect/fns/process_group.py (hunk 1 above)
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: Oh, I don't like this doesn't fail already in unit tests... I will check it.

Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: @Chr96er Done. Added the check in the unit tests and fixed the gap creation.

File: src/pipe_gaps/pipelines/detect/fns/process_group.py [hunk 2]
```diff
@@ -70,6 +71,24 @@ def process(
...
+
+            # Emit open version if daily mode would have created one on any day between OFF and ON.
+            # This ensures range processing produces the same table state as daily processing,
+            # so that a subsequent reprocess can always reconstruct the gap from the open version.
+            off_m = self._gd.off_message_from_gap(gap)
+            off_date = datetime_from_timestamp(off_m[self.KEY_TIMESTAMP]).date()
+
+            should_emit_open = any(
+                self._gd.eval_open_gap(off_m, off_date + timedelta(days=i))
+                for i in range(ceil(gap["duration_h"] / 24) - 1)  # exclude the ON day
```
[the remark this answers is not in the document]
Comment (Chr96er) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: @tomaslink any reason you marked this as resolved? I just ran the gaps integration tests and it looks like they're failing exactly because of that. Isn't the issue real that it's not at all about the gap duration but whether or not the start and end are on the same calendar day?

[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: @Chr96er I don't believe this is an issue. The example copilot gives should be covered by the current code. Let's discuss in the meeting.

File: src/pipe_gaps/pipelines/detect/fns/process_group.py [hunk 3]
```diff
@@ -70,6 +71,24 @@ def process(
 
         for gap in gaps:
             self._debug_gap(gap)
+
+            # Emit open version if daily mode would have created one on any day between OFF and ON.
+            # This ensures range processing produces the same table state as daily processing,
+            # so that a subsequent reprocess can always reconstruct the gap from the open version.
+            off_m = self._gd.off_message_from_gap(gap)
```
[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py:77]: @Chr96er We don't use unnormalized version in production. We just need to deprecate that option. I wouldn't pollute the code with this...

File: src/pipe_gaps/pipelines/detect/fns/process_group.py [hunk 4]
```diff
@@ -70,6 +71,24 @@ def process(
...
+            should_emit_open = any(
+                self._gd.eval_open_gap(off_m, off_date + timedelta(days=i))
+                for i in range(ceil(gap["duration_h"] / 24) - 1)  # exclude the ON day
+            )
+
+            if should_emit_open:
+                logger.debug("Emitting open gap for recovery...")
+                open_gap = self._gd.create_gap(off_m=off_m, gap_id=gap[self.KEY_GAP_ID])
+                self._debug_gap(open_gap)
+                yield open_gap
```
[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: @Chr96er I don't believe this is the case, because the gap detected in this part of the code is not a gap that was previously open, it is detecting the OFF and ON simultaneously.

File: src/pipe_gaps/pipelines/detect/fns/process_boundaries.py [hunk 5]
```diff
@@ -99,6 +100,24 @@ def process(
...
+                    self._gap_detector.eval_open_gap(off_m, off_date + timedelta(days=i))
+                    for i in range(ceil(g["duration_h"] / 24) - 1)  # exclude the ON day)
+                )
+
+                if should_emit_open:
+                    logger.debug("Emitting open gap for recovery...")
+                    gap_id = g[self.KEY_GAP_ID]
+                    open_gap = self._gap_detector.create_gap(off_m=off_m, gap_id=gap_id)
+                    self._debug_gap(open_gap)
+                    gaps[f"{gap_id}_open"] = open_gap  # temporary key to avoid overwriting
```
[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_boundaries.py]: @Chr96er I don't believe this is true, since the open gaps are treated in the step one of that part of the code. In this step two, we skip with:

```python
if not self._is_message_in_range(left.last_message()):
    # Otherwise should be an open gap and we handle those in step one.
    continue
```

So this branch should be treating gaps that never had an open gap before...

File: src/pipe_gaps/pipelines/detect/fns/process_group.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: @Chr96er Ok, bug understood and fixed.

File: src/pipe_gaps/core/gap_detector.py [hunk 6]
```diff
@@ -333,7 +334,7 @@ def create_gap(
         gap.update({
             self.KEY_GAP_ID: gap_id,
             self.KEY_SSVID: ssvid,
-            self.KEY_VERSION: int(datetime.now(tz=timezone.utc).timestamp()),
+            self.KEY_VERSION: int(version),
```
Comment (tomaslink) [inline src/pipe_gaps/core/gap_detector.py:337]: @Chr96er This change was to make the version stable, so for the same input getting always the same output. Are you ok with this change? We shouldn't have more than 2 versions of a gap anyway, at the most 1 open gap and one closed gap. So should be enough to order them. But I would like to know your opinion. 

CC: @andres-arana

Comment (Chr96er) [reply src/pipe_gaps/core/gap_detector.py:337]: I like the idea and _think_ that should work. Sounds like a big and potentially risky change but I think running integration tests should catch any potential issues.

File: src/pipe_gaps/pipelines/detect/fns/process_boundaries.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_boundaries.py]: I found some issues related to duplicate open gaps. I would like to resolve them in another PR.

File: src/pipe_gaps/pipelines/detect/fns/process_group.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomaslink) [reply src/pipe_gaps/pipelines/detect/fns/process_group.py]: I found some issues related to duplicate open gaps. I would like to resolve them in another PR.
